### ISSUE TITLE
Hint not clearing properly

### ISSUE THREAD

**Issue by williamboman** [erming/tabcomplete#14]
When programmatically emptying the input field the hint is not cleared, causing this issue; ![](https://cloud.githubusercontent.com/assets/556268/5360161/d6fdd08c-7fc2-11e4-9836-74c796ae539f.png)

Here's a GIF demonstrating the issue.
![](http://i.imgur.com/hQFwHRC.gif)
1. When typing something that does not exist in the filter array, pressing `x` (clearing) works as expected.
2. When typing something that does exist in the filter array, pressing `x` does not hide the hint. From reading through the code quickly I'd assume this is because it only clears the hint on `backspace` keydowns.
3. Hitting backspace to manually remove works as expected.

This is the on-click event for the `x` button

```
$clear.click(function(e) {
  e.preventDefault();
  $filter_by.val('').trigger('keyup').focus();
});
```

I've also tried

```
$clear.on('click', function(e) {
  e.preventDefault();
  var customEvent = $.Event('keydown');
  customEvent.which = 8 // Backspace key
  $filter_by.val('').trigger(customEvent).focus();
});
```

without success.

@erming, do you have any ideas how to solve this? I don't have time right now to properly read through this lib and do a PR :(.

**Comment by erming** [comment]
Hey!

Yeah, I got an idea how to solve this.

I've done something like this before: I will hook `.val()` to emit an event, which I can then listen on, and then something like this: `input.on("val", function() { .. update hint .. });`

I'll ping you when I'm done (later tonight), I Just need to head down to the grocery store before it closes.

**Comment by erming** [comment]
Holy.. That was horrible!

Just spent over an hour trying to make the hint to update when you `.val(str)`. Well. It kind of worked, but `.val()` moves the input caret to the beginning of the input. Took me a long while to figure out why the hints didn't trigger. And when I tried to fix that, the input focus messed up, sometimes moving focus to the hint input element instead. Argh!

Anyway, while I couldn't implement it natively without tearing my hair off, here's your solution:

``` javascript
var input = $("#input");
input.val("a");
input[0].selectionStart = input.val().length; // Move caret to end of your string
input.trigger("input"); // Trigger hint
```

Hope it works for you. Good luck!

**P.S:**
I just realized that you only needed to clear the input, not trigger hints via `.val()`. Try this:

``` javascript
$("#input").val("").trigger("input");
```
